src/effects/effects_base/internal/effectsprovider.cpp

```diff
@@ -182,3 +187,241 @@ muse::Ret EffectsProvider::performEffect(AudacityProject& project, Effect* effec
 
     return success ? muse::make_ok() : muse::make_ret(Ret::Code::UnknownError);
 }
+
+static TransportSequences MakeTransportTracks(
+    TrackList& trackList, bool selectedOnly, bool nonWaveToo = false)
+{
+    TransportSequences result;
+    {
+        const auto range = trackList.Any<WaveTrack>()
+                           + (selectedOnly ? &Track::IsSelected : &Track::Any);
+        for (auto pTrack : range) {
+            result.playbackSequences.push_back(
+                StretchingSequence::Create(*pTrack, pTrack->GetClipInterfaces()));
+        }
+    }
+    if (nonWaveToo) {
+        const auto range = trackList.Any<const PlayableTrack>()
+                           + (selectedOnly ? &Track::IsSelected : &Track::Any);
+        for (auto pTrack : range) {
+            if (!track_cast<const SampleTrack*>(pTrack)) {
+                if (auto pSequence
+                        =std::dynamic_pointer_cast<const OtherPlayableSequence>(
+                              pTrack->shared_from_this())
+                        ) {
+                    result.otherPlayableSequences.push_back(pSequence);
+                }
+            }
+        }
+    }
+    return result;
+}
+
+static void EffectPreview(EffectBase& effect,
+                          EffectSettingsAccess& access, std::function<void()> updateUI, bool dryOnly)
+{
+    auto cleanup0 = effect.BeginPreview(access.Get());
+
+    // These are temporary state in the Effect object that are meant be moved to
+    // a new class EffectContext
+    const auto numTracks = effect.mNumTracks;
+    const auto rate = effect.mProjectRate;
+    const auto& factory = effect.mFactory;
+    auto& mT0 = effect.mT0;
+    auto& mT1 = effect.mT1;
+    auto& mTracks = effect.mTracks;
+    auto& mProgress = effect.mProgress;
+    auto& mIsPreview = effect.mIsPreview;
+
+    // Get certain immutable properties of the effect
+    const auto previewFullSelection = effect.PreviewsFullSelection();
+    const auto isLinearEffect = effect.IsLinearEffect();
+
+    if (numTracks == 0) { // nothing to preview
+        return;
+    }
+
+    auto gAudioIO = AudioIO::Get();
+    if (gAudioIO->IsBusy()) {
+        return;
+    }
+
+    const auto FocusDialog = BasicUI::FindFocus();
+    assert(FocusDialog); // postcondition
+
+    double previewDuration;
+    bool isNyquist = effect.GetFamily() == NYQUISTEFFECTS_FAMILY;
+    bool isGenerator = effect.GetType() == EffectTypeGenerate;
+
+    // Mix a few seconds of audio from all of the tracks
+    double previewLen;
+    gPrefs->Read(wxT("/AudioIO/EffectsPreviewLen"), &previewLen, 6.0);
```

Comment:
The value 6.0 appears to be a magic number

Reply:
https://github.com/audacity/audacity/pull/7537#discussion_r1808779273